core: Replace THIS->ctx with global_ctx

mohit84 opened the pull request:
In the patch https://github.com/gluster/glusterfs/pull/1614
we have introduced a global_ctx to access a ctx variable.
With the help of this global_ctx variable now it is not
require to call THIS to access ctx and avoid unnecessary
resource consumption to call THIS.

Fixes: #1874
Change-Id: I1ceb193dddf92754f20905379c92ac7b8f89e1fb
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 committed:
core: Replace THIS->ctx with global_ctx

In the patch https://github.com/gluster/glusterfs/pull/1614
we have introduced a global_ctx to access a ctx variable.
With the help of this global_ctx variable now it is not
require to call THIS to access ctx and avoid unnecessary
resource consumption to call THIS.

Fixes: #1874
Change-Id: I1ceb193dddf92754f20905379c92ac7b8f89e1fb
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

mykaul commented on xlators/storage/posix/src/posix-helpers.c:
```diff
@@ -2155,7 +2155,7 @@ posix_health_check_thread_proc(void *data)
     xlator_list_t **trav_p = NULL;
     int count = 0;
     gf_boolean_t victim_found = _gf_false;
-    glusterfs_ctx_t *ctx = THIS->ctx;
```
where do we use this? We only use it in the abort flow - we can initialize it only there.

mohit84 replied:
done

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run brick-mux regression

xhernandez commented on libglusterfs/src/dict.c:
```diff
@@ -54,8 +54,6 @@ get_new_data()
 
     if (global_ctx) {
```
Can `global_ctx` be NULL ?

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -118,7 +118,7 @@ xlator_instantiate_va(const char *type, const char *format, va_list arg)
     CDS_INIT_LIST_HEAD(&xl->volume_options);
 
-    xl->ctx = this->ctx;
+    xl->ctx = global_ctx;
```
Do we really need this ? I think we should remove the `ctx` field from `xlator_t`.

xhernandez commented on xlators/storage/posix/src/posix-helpers.c:
```diff
@@ -2215,6 +2215,7 @@ posix_health_check_thread_proc(void *data)
        count is more than 1 it means brick mux has enabled
     */
+    ctx = global_ctx;
     if (this->ctx->active) {
         top = this->ctx->active->first;
```
```suggestion
    if (ctx->active) {
        top = ctx->active->first;
```

xhernandez reviewed (commented):
There are many places where we copy `global_ctx` to a local variable before using it. Is it really necessary now ?

vh05 committed:
features/shard: unlink fails due to nospace to mknod marker file

When we hit the max capacity of the storage space, shard_unlink()
starts failing if there is no space left on the brick to create a
marker file.

shard_unlink() happens in below steps:

1. create a marker file in the name of gfid of the base file under
BRICK_PATH/.shard/.remove_me
2. unlink the base file
3. shard_delete_shards() deletes the shards in background by
picking the entries in BRICK_PATH/.shard/.remove_me

If a marker file creation fails then we can't really delete the
shards which eventually a problem for user who is looking to make
space by deleting unwanted data.

Solution:
Create the marker file by marking xdata = GLUSTERFS_INTERNAL_FOP_KEY
which is considered to be internal op and allowed to create under
reserved space.

Fixes: #2038
Change-Id: I7facebab940f9aeee81d489df429e00ef4fb7c5d
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

mohit84 commented:
/run regression

xhernandez commented on api/src/glfs-mgmt.c:
```diff
@@ -59,13 +57,13 @@ glfs_process_volfp(struct glfs *fs, FILE *fp)
     }
 
-    ret = glusterfs_graph_prepare(graph, ctx, fs->volname);
+    ret = glusterfs_graph_prepare(graph, global_ctx, fs->volname);
```
I wouldn't send `global_ctx` as an argument. If the function needs it, it can get it directly from the global variable.

xhernandez commented on api/src/glfs-mgmt.c:
```diff
@@ -718,12 +690,12 @@ glfs_mgmt_getspec_cbk(struct rpc_req *req, struct iovec *iov, int count,
     }
 
-    if (ret && ctx && !ctx->active) {
+    if (ret && global_ctx && !global_ctx->active) {
```
```suggestion
    if (ret && !global_ctx->active) {
```

xhernandez commented on api/src/glfs.c:
```diff
@@ -931,7 +919,7 @@ priv_glfs_new_from_ctx(glusterfs_ctx_t *ctx)
         goto out;
 
-    fs->ctx = ctx;
+    global_ctx = ctx;
```
`ctx` shouldn't be required by this function as it's global. However it's part of a public API, so it can't be removed easily. I would simply ignore it (note that it should already be equal to `global_ctx`). In any case we must not update `global_ctx` here.

xhernandez commented on api/src/glfs.c:
```diff
@@ -1103,7 +1091,7 @@ glfs_init_async(struct glfs *fs, glfs_init_cbk cbk)
     int ret = -1;
 
-    if (!fs || !fs->ctx) {
+    if (!fs || !global_ctx) {
```
```suggestion
    if (!fs) {
```

xhernandez commented on api/src/glfs.c:
```diff
@@ -1124,7 +1112,7 @@ pub_glfs_init(struct glfs *fs)
     DECLARE_OLD_THIS;
 
-    if (!fs || !fs->ctx) {
+    if (!fs || !global_ctx) {
```
```suggestion
    if (!fs) {
```

xhernandez commented on api/src/glfs.c:
```diff
@@ -1596,21 +1576,18 @@ GFAPI_SYMVER_PUBLIC_DEFAULT(glfs_sysrq, 3.10.0)
     }; /* should not exceed 1024 chars */
 
-    if (!fs || !fs->ctx) {
+    if (!fs || !global_ctx) {
```
```suggestion
    if (!fs) {
```

xhernandez commented on glusterfsd/src/glusterfsd-mgmt.c:
```diff
@@ -1362,19 +1335,8 @@ glusterfs_handle_brick_status(rpcsvc_request_t *req)
-    }
-    active = ctx->active;
-    if (ctx->active->first == NULL) {
+    active = global_ctx->active;
```
Is it not necessary to check if `active` is NULL ?

xhernandez commented on glusterfsd/src/glusterfsd-mgmt.c:
```diff
@@ -2364,12 +2312,12 @@ mgmt_getspec_cbk(struct rpc_req *req, struct iovec *iov, int count,
     }
 
-    if (ret && ctx && !ctx->active) {
+    if (ret && global_ctx && !global_ctx->active) {
```
```suggestion
    if (ret && !global_ctx->active) {
```

xhernandez commented on glusterfsd/src/glusterfsd-mgmt.c:
```diff
@@ -1362,19 +1335,8 @@ glusterfs_handle_brick_status(rpcsvc_request_t *req)
         goto out;
     }
 
-    ctx = glusterfsd_ctx;
```
The declaration of `glusterfsd_ctx` should also be removed.

xhernandez commented on libglusterfs/src/defaults-tmpl.c:
```diff
@@ -156,8 +156,8 @@ default_notify(xlator_t *this, int32_t event, void *data, ...)
              */
-            if (!parent && this->ctx && this->ctx->root) {
-                xlator_notify(this->ctx->root, event, this->graph, NULL);
+            if (!parent && global_ctx && global_ctx->root) {
```
```suggestion
            if (!parent) {
```

xhernandez commented on libglusterfs/src/defaults-tmpl.c:
```diff
@@ -167,7 +167,7 @@ default_notify(xlator_t *this, int32_t event, void *data, ...)
 
             if (event == GF_EVENT_CHILD_DOWN &&
-                !(this->ctx && this->ctx->root) && (graph->top == this)) {
+                !(global_ctx && global_ctx->root) && (graph->top == this)) {
```
```suggestion
                && (graph->top == this)) {
```

xhernandez commented on libglusterfs/src/defaults-tmpl.c:
```diff
@@ -183,8 +183,8 @@ default_notify(xlator_t *this, int32_t event, void *data, ...)
 
-            if (!parent && this->ctx && this->ctx->root)
-                xlator_notify(this->ctx->root, event, data, NULL);
+            if (!parent && global_ctx && global_ctx->root)
```
```suggestion
            if (!parent)
```

xhernandez commented on libglusterfs/src/dict.c:
```diff
@@ -1475,16 +1461,7 @@ dict_reset(dict_t *dict)
-    dict->members_list = NULL;
-
-    UNLOCK(&dict->lock);
+    dict_foreach(dict, dict_remove_foreach_fn, NULL);
```
This is inefficient. It was recently changed to make it faster. See #2011. Please keep those changes.

xhernandez commented on libglusterfs/src/events.c:
```diff
@@ -36,35 +36,31 @@ _gf_event(eventtypes_t event, const char *fmt, ...)
 
-    if (ctx) {
-        volfile_server_transport = ctx->cmd_args.volfile_server_transport;
+    if (global_ctx) {
```
Not necessary.

xhernandez commented on libglusterfs/src/events.c:
```diff
@@ -36,35 +36,31 @@ _gf_event(eventtypes_t event, const char *fmt, ...)
 
     /* host = NULL returns localhost */
-    if (ctx && ctx->cmd_args.volfile_server &&
+    if (global_ctx && global_ctx->cmd_args.volfile_server &&
```
```suggestion
    if (global_ctx->cmd_args.volfile_server &&
```

xhernandez commented on libglusterfs/src/glusterfs/stack.h:
```diff
@@ -34,6 +34,7 @@ typedef struct call_pool call_pool_t;
 #include "glusterfs/libglusterfs-messages.h"
 #include "glusterfs/timespec.h"
 
+extern glusterfs_ctx_t *global_ctx;
```
Is `stack.h` the best place to declare this ? probably `globals.h` or even `common-utils.h` would be a better place.

xhernandez commented on libglusterfs/src/logging.c:
```diff
@@ -79,31 +79,31 @@ static char gf_level_strings[] = {
-    if (THIS->ctx) {
-        THIS->ctx->log.logrotate = 1;
-        THIS->ctx->log.cmd_history_logrotate = 1;
+    if (global_ctx) {
```
Is it necessary ?

There are other similar checks in this file.

xhernandez commented on libglusterfs/src/logging.c:
```diff
@@ -1920,18 +1883,11 @@ _gf_msg(const char *domain, const char *file, const char *function,
-        return -1;
-
-    ctx = this->ctx;
-    if (ctx == NULL) {
```
I've seen this could happen. If there are error during initialization of the `global_ctx`, messages can be sent before `global_ctx` is valid.

xhernandez commented on libglusterfs/src/mem-pool.c:
```diff
@@ -129,7 +129,7 @@ gf_mem_acct_enabled(void)
     /* Low-level __gf_xxx() may be called
        before ctx is initialized. */
-    return x->ctx && x->ctx->mem_acct_enable;
+    return global_ctx && global_ctx->mem_acct_enable;
```
```suggestion
    return global_ctx->mem_acct_enable;
```

xhernandez commented on libglusterfs/src/xlator.c:
```diff
@@ -724,10 +724,10 @@ xlator_mem_acct_init(xlator_t *xl, int num_types)
         return -1;
 
-    if (!xl->ctx)
+    if (!global_ctx)
```
Shouldn't happen.

xhernandez commented on libglusterfs/src/xlator.c:
```diff
@@ -824,7 +824,7 @@ xlator_members_free(xlator_t *xl)
     GF_FREE(xl->name);
     GF_FREE(xl->type);
-    if (!(xl->ctx && xl->ctx->cmd_args.vgtool != _gf_none) && xl->dlhandle)
+    if (!(global_ctx && global_ctx->cmd_args.vgtool != _gf_none) && xl->dlhandle)
```
```suggestion
    if (global_ctx->cmd_args.vgtool == _gf_none && xl->dlhandle)
```

xhernandez commented on libglusterfs/src/xlator.c:
```diff
@@ -951,19 +951,16 @@ xlator_mem_cleanup(xlator_t *this)
     gf_boolean_t graph_cleanup = _gf_false;
 
-    if (this->call_cleanup || !this->ctx)
+    if (this->call_cleanup || !global_ctx)
```
```suggestion
    if (this->call_cleanup)
```

xhernandez commented on rpc/rpc-transport/socket/src/socket.c:
```diff
@@ -3118,7 +3098,7 @@ socket_server_event_handler(int fd, int idx, int gen, void *data, int poll_in,
         if (new_sockaddr.ss_family != AF_UNIX)
             new_trans->options = dict_ref(this->options);
-        new_trans->ctx = this->ctx;
+        new_trans->ctx = global_ctx;
```
`ctx` field of `rpc_transport_t` should be removed.

xhernandez commented on xlators/cluster/dht/src/dht-rebalance.c:
```diff
@@ -4431,8 +4423,8 @@ gf_defrag_start_crawl(void *data)
         gf_defrag_status_get(conf, status, _gf_true);
-        if (ctx && ctx->notify)
-            ctx->notify(GF_EN_DEFRAG_STATUS, status);
+        if (global_ctx && global_ctx->notify)
```
Is it possible that this condition could be false ?

xhernandez commented on xlators/features/quota/src/quotad-helpers.c:
```diff
@@ -36,7 +36,7 @@ get_quotad_aggregator_state(xlator_t *this, rpcsvc_request_t *req)
     state->itable = active_subvol->itable;
 
-    state->pool = this->ctx->pool;
+    state->pool = global_ctx->pool;
```
Do we need to save the global pool in a private structure ? we could use the global one where required.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-locks.c:
```diff
@@ -588,7 +588,7 @@ glusterd_mgmt_v3_lock(const char *name, uuid_t uuid, uint32_t *op_errno,
 
-    mgmt_lock_timer_ctx = mgmt_lock_timer_xl->ctx;
+    mgmt_lock_timer_ctx = global_ctx;
     if (!mgmt_lock_timer_ctx) {
```
Needed ?

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-locks.c:
```diff
@@ -681,7 +681,7 @@ gd_mgmt_v3_unlock_timer_cbk(void *data)
 
-        mgmt_lock_timer_ctx = mgmt_lock_timer_xl->ctx;
+        mgmt_lock_timer_ctx = global_ctx;
         GF_VALIDATE_OR_GOTO(this->name, mgmt_lock_timer_ctx, ret_function);
```
Needed ?

xhernandez commented on xlators/mgmt/glusterd/src/glusterd-locks.c:
```diff
@@ -804,7 +804,7 @@ glusterd_mgmt_v3_unlock(const char *name, uuid_t uuid, char *type)
 
-        mgmt_lock_timer_ctx = mgmt_lock_timer_xl->ctx;
+        mgmt_lock_timer_ctx = global_ctx;
         GF_VALIDATE_OR_GOTO(this->name, mgmt_lock_timer_ctx, out);
```
Needed ?

xhernandez commented on xlators/mount/fuse/src/fuse-helpers.c:
```diff
@@ -130,7 +130,7 @@ get_fuse_state(xlator_t *this, fuse_in_header_t *finh)
     state->itable = active_subvol->itable;
 
-    state->pool = this->ctx->pool;
+    state->pool = global_ctx->pool;
```
Do we need to save the pool in a private structure ?

xhernandez commented on xlators/performance/read-ahead/src/read-ahead.c:
```diff
@@ -1126,7 +1126,7 @@ init(xlator_t *this)
     }
 
-    conf->page_size = this->ctx->page_size;
+    conf->page_size = global_ctx->page_size;
```
Needed ?

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run full regression

mohit84 commented:
/run full regression
